src/lib.rs

```diff
@@ -0,0 +1,360 @@
+        self.http_client = Some(v);
+        self
+    }
+
+    /// Use this method to provide the application callback that is used to
+    /// build the attestation evidence for this verification session.  See
+    /// EvidenceCreationCb.
+    pub fn with_evidence_creation_cb(mut self, v: EvidenceCreationCb) -> ChallengeResponseBuilder {
+        self.evidence_creation_cb = Some(v);
+        self
+    }
+
+    /// Instantiate a valid ChallengeResponse object, or fail with an error.
+    pub fn build(self) -> Result<ChallengeResponse, Error> {
+        let base_url_str = self
+            .base_url
+            .ok_or(Error::ConfigError("missing API endpoint".to_string()))?;
+
+        Ok(ChallengeResponse {
+            base_url: url::Url::parse(&base_url_str)
+                .map_err(|e| Error::ConfigError(e.to_string()))?,
+            http_client: self
+                .http_client
+                .ok_or(Error::ConfigError("missing HTTP client".to_string()))?,
+            evidence_creation_cb: self.evidence_creation_cb.ok_or(Error::ConfigError(
+                "missing evidence creation callback".to_string(),
+            ))?,
+        })
+    }
+}
+
+/// The object on which one or more challenge-response verification sessions can
+/// be run.  Always use the ChallengeResponseBuilder to instantiate it.
+pub struct ChallengeResponse {
+    base_url: url::Url,
+    http_client: reqwest::blocking::Client,
+    evidence_creation_cb: EvidenceCreationCb,
+}
+
+/// Nonce configuration: either the size (Size) of the nonce generated by the
+/// server (use 0 to let the server also pick the size of the challenge), or an
+/// explicit nonce (Value) supplied as a byte array.
+pub enum Nonce {
+    Size(usize),
+    Value(Vec<u8>),
+}
+
+impl ChallengeResponse {
+    /// Run a challenge-response verification session using the supplied nonce
+    /// configuration.  Returns the raw attestation results, or an error on
+    /// failure.
+    pub fn run(&self, nonce: Nonce) -> Result<String, Error> {
+        // create new c/r verification session on the veraison side
+        //let (session_url, session) = self.new_session_with_nonce(nonce)?;
+        let (session_url, session) = self.new_session(&nonce)?;
+
+        // invoke the user-provided evidence builder callback with per-session parameters
+        let (evidence, media_type) =
+            (self.evidence_creation_cb)(session.nonce(), session.accept())?;
+
+        // send evidence for verification to the session endpoint
+        let attestation_result = self.challenge_response(&evidence, &media_type, &session_url)?;
+
+        // return veraison's attestation results
+        Ok(attestation_result)
+    }
+
+    fn new_session(&self, nonce: &Nonce) -> Result<(String, ChallengeResponseSession), Error> {
+        // ask veraison for a new session object
+        let resp = self
+            .new_session_request(nonce)
+            .map_err(|e| Error::ApiError(e.to_string()))?;
```

Comment:
Is this mapping needed? Seems like `new_session_request` already returns the correct type of error.

Reply:
good catch!

Reply:
fixed in [be6ebbb](https://github.com/veraison/rust-apiclient/pull/4/commits/be6ebbb506f46bd02b947493629f4852ec36dae0)